mmdet/models/detectors/base_detr.py

```diff
@@ -0,0 +1,327 @@
+    """Base class for Transformer-based detectors.
+
+    Transformer-based detectors use an encoder to process output features of
+    neck, then several queries interactive with the output features of encoder
+    and do the regression and classification.
+
+    Args:
+        backbone (:obj:`ConfigDict` or dict): Config of backbone.
+        neck (:obj:`ConfigDict` or dict): Config of neck.
+            Defaults to None.
+        encoder_cfg (:obj:`ConfigDict` or dict): Config of encoder.
+            Defaults to None.
+        decoder_cfg (:obj:`ConfigDict` or dict): Config of decoder.
+            Defaults to None.
+        positional_encoding_cfg (:obj:`ConfigDict` or dict): Config of
+            positional encoding. Defaults to None.
+        bbox_head (:obj:`ConfigDict` or dict): Config for position
+            encoding. Defaults to None.
+        num_query (int): Number of decoder query in Transformer.
+            Defaults to 100.
+        train_cfg (:obj:`ConfigDict` or dict): Training config of transformer
+            head. Defaults to None.
+        test_cfg (:obj:`ConfigDict` or dict): Testing config of transformer
+            head. Defaults to None.
+        data_preprocessor (dict or ConfigDict, optional): The pre-process
+           config of :class:`BaseDataPreprocessor`.  it usually includes,
+            ``pad_size_divisor``, ``pad_value``, ``mean`` and ``std``.
+            Defaults to None.
+        init_cfg (:obj:`ConfigDict` or dict or list[:obj:`ConfigDict` or \
+            dict], optional): Initialization config dict. Defaults to None.
+    """
+
+    def __init__(self,
+                 backbone: ConfigType,
+                 neck: OptConfigType = None,
+                 encoder_cfg: OptConfigType = None,
+                 decoder_cfg: OptConfigType = None,
+                 positional_encoding_cfg: OptConfigType = None,
+                 bbox_head: OptConfigType = None,
+                 num_query: int = 100,
+                 train_cfg: OptConfigType = None,
+                 test_cfg: OptConfigType = None,
+                 data_preprocessor: OptConfigType = None,
+                 init_cfg: OptMultiConfig = None) -> None:
+        super().__init__(
+            data_preprocessor=data_preprocessor, init_cfg=init_cfg)
+        # process args
+        bbox_head.update(train_cfg=train_cfg)
+        bbox_head.update(test_cfg=test_cfg)
+        self.train_cfg = train_cfg
+        self.test_cfg = test_cfg
+        self.encoder_cfg = encoder_cfg
+        self.decoder_cfg = decoder_cfg
+        self.positional_encoding_cfg = positional_encoding_cfg
+        self.num_query = num_query
+
+        # init model layers
+        self.backbone = MODELS.build(backbone)
+        if neck is not None:
+            self.neck = MODELS.build(neck)
+        self.bbox_head = MODELS.build(bbox_head)
+        self._init_layers()
+
+    @abstractmethod
+    def _init_layers(self) -> None:
```

Comment:
This method and its name are strange. The backbone, neck, and head are all layers, why not build them in this method?

Reply:
Hi, thanks for your review !!! 

In `_init_layers()` we initialize all layers that don't need the registry mechanism, like `encoder`, `decoder`, etc.

We have thought about other names like `_init_transformer()`. But we found that this function often needs to initialize other layers, such as `pos_trans_fc`, `memory_trans_fc`, `level_embed`.

We also thought about whether we could put the initialization of `backbone`, `neck`, and `bbox_head` into this part, but since `_init_layers()` has no args, this would cause some redundancy as follows:

```Python
def __init__(...) -> None:
    ... 
    self.backbone = backbone  # OptConfigType
    if neck is not None:
        self.neck = neck  # OptConfigType
    self.bbox_head = bbox_head  # OptConfigType

    self._init_layers()
    ...

@abstractmethod
def _init_layers(self) -> None:
    """Initialize layers."""
    self.backbone = MODELS.build(self.backbone)
    if self.with_neck:
        self.neck = MODELS.build(self.neck)
    self.bbox_head = MODELS.build(bbox_head)
    ...
```

We will discuss the scheme about `_init_layer()` at our next meeting. Looking forward to feedback~